src/torrent.cpp

```diff
@@ -5291,12 +5291,20 @@ namespace libtorrent
 		TORRENT_ASSERT(std::count(m_peers_to_disconnect.begin()
 			, m_peers_to_disconnect.end(), p) == 0);
 
-		std::weak_ptr<torrent> weak_t = shared_from_this();
-		m_peers_to_disconnect.push_back(p);
-		m_deferred_disconnect.post(m_ses.get_io_service(), [=]() {
-			std::shared_ptr<torrent> t = weak_t.lock();
-			if (t) t->on_remove_peers();
-		});
+		// we only schedule the peer for actual removal if in fact
+		// it is attached to this torrent, since this method can
+		// be called from the attach_peer path and fail to do so
+		// because of too many connections
+		if (p->associated_torrent().lock().get() == this)
```

Comment:
I see. really the property to test for here is whether this peer connection will be kept alive until the deferred function is called or not. could you make it clearer in the comment that before a peer_connection is attached to a torrent, the session_impl object may remove it at any time, which may be while the non-owning pointer is waiting for the deferred function to be called.

I have a feeling that there's a simpler way to express this patch. I have two ideas:

1. Instead of removing the peer from ``m_connections`` in attach_peer(), could you do that in an else block to this if-statement? If that would work, that would seems a bit more robust.

2. Perhaps ``m_peers_to_disconnect`` should hold shared_ptr to the peers